src/shogun/machine/Pipeline.cpp

```diff
@@ -0,0 +1,149 @@
+/*
+ * This software is distributed under BSD 3-clause license (see LICENSE file).
+ *
+ * Authors: Wuwei Lin
+ */
+
+#include <shogun/machine/Pipeline.h>
+
+namespace shogun
+{
+	CPipeline::CPipeline() : CMachine()
+	{
+	}
+
+	CPipeline::~CPipeline()
+	{
+		for (auto&& stage : m_stages)
+		{
+			visit([](auto&& object) { SG_UNREF(object) }, stage);
+		}
+	}
+
+	CPipeline* CPipeline::with(CTransformer* transformer)
+	{
+		REQUIRE(
+		    m_stages.empty() ||
+		        holds_alternative<CTransformer*>(m_stages.back()),
+		    "Transformers can not be placed after machines.\n");
```

Comment:
Could you maybe make this "Last element in pipeline is %s", and then put the machine name there?